llvm-external-projects/iree-compiler-api/python/iree/compiler/tools/binaries.py

```diff
@@ -0,0 +1,283 @@
+# Lint-as: python3
+"""Utilities for locating and invoking compiler tool binaries."""
+
+# Copyright 2020 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import importlib
+import io
+import os
+import platform
+import subprocess
+import sys
+import textwrap
+import threading
+
+from typing import List, Optional, Union
+
+__all__ = [
+    "find_tool",
+    "invoke_immediate",
+    "invoke_pipeline",
+    "get_tool_path",
+    "CompilerToolError",
+]
+
+_BUILTIN_TOOLS = [
+    "ireec",
+]
+
+# In normal distribution circumstances, each named tool is associated with
+# a python module that provides a `get_tool` function for getting its absolute
+# path. This dictionary maps the tool name to the module.
+_TOOL_MODULE_MAP = {
+    "iree-import-tflite": "iree.tools.tflite",
+    "iree-import-xla": "iree.tools.xla",
+    "iree-import-tf": "iree.tools.tf",
+}
+
+# Map of tool module to package name as distributed to archives (used for
+# error messages).
+_TOOL_MODULE_PACKAGES = {
+    "iree.tools.tf": "iree-tools-tf",
+    "iree.tools.tflite": "iree-tools-tflite",
+    "iree.tools.xla": "iree-tools-xla",
+}
+
+# Environment variable holding directories to be searched for named tools.
+# Delimitted by os.pathsep.
+_TOOL_PATH_ENVVAR = "IREE_TOOL_PATH"
+
+
+class CompilerToolError(Exception):
+  """Compiler exception that preserves the command line and error output."""
+
+  def __init__(self, process: subprocess.CompletedProcess):
+    try:
+      errs = process.stderr.decode("utf-8")
+    except:
+      errs = str(process.stderr)  # Decode error or other: best we can do.
+
+    tool_name = os.path.basename(process.args[0])
+    super().__init__(f"Error invoking IREE compiler tool {tool_name}\n"
+                     f"Diagnostics:\n{errs}\n\n"
+                     f"Invoked with:\n  {' '.join(process.args)}")
```

Comment:
Does `process.args` include `tool_name`?

Some compiler errors include large IR dumps, which can fill console output buffers. Having context about the tool + args printed last can help there. So one suggestion here: how about including the tool name twice?

```
Error invoking IREE compiler tool iree-import-tf
Diagnostics:
  foo
Invoked with:
  iree-import-tf --arg1 --arg2
```

then the invocation could also be copy-pasted directly